lib/src/refs.rs

```diff
@@ -108,6 +110,27 @@ pub fn merge_ref_targets(
     }
 }
 
+pub fn merge_remote_refs(
+    index: &dyn Index,
+    left: &RemoteRef,
+    base: &RemoteRef,
+    right: &RemoteRef,
+) -> RemoteRef {
+    // Just merge target and state fields separately. Strictly speaking, merging
+    // target-only change and state-only change shouldn't automatically mark the
+    // new target as tracking. However, many faulty merges will end up in local
+    // or remote target conflicts (since fast-forwardable move can be safely
+    // "tracked"), and the conflicts will require user intervention anyway. So
+    // there wouldn't be much reason to handle these merges precisely.
+    let target = merge_ref_targets(index, &left.target, &base.target, &right.target);
+    // If state conflicted, pick the lower (i.e. non-tracking) state. The choice
+    // is arbitrary, but it should be slightly easier to re-track untracked refs
+    // than recover from false tracking state.
```

Comment:
I don't think there can be a conflict in practice right now because we only have two states, and doing a 3-way merge of two states will never conflict. Hmm, if the branch was added on both sides (e.g. by concurrent operations) and one side added it as tracking and the other as non-tracking, I suppose we'll resolve that as tracking because the absent ref is considered non-tracking. That's probably very rare in practice and it seems like a reasonable automatic resolution anyway.

Reply:
True. Updated the comment and simplified the unwrapping a bit.